src/WCSimAmBeGen.cc

```diff
@@ -0,0 +1,127 @@
+G4double WCSimAmBeGen::gammaProbabilities[3] = {0.26, 0.65, 0.08};
+G4double WCSimAmBeGen::gammaEnergies[3] = {0.0, 4.4, 7.7};
+G4int    WCSimAmBeGen::pdgids[2] = {2112, 22};
+
+WCSimAmBeGen::WCSimAmBeGen(){
+  // Initialise
+  wcsimdir = string(getenv("WCSIMDIR"))+"data/";
+
+  gs_path = wcsimdir + "ground_state_spectrum.txt";
+  fe_path = wcsimdir + "first_excited_spectrum.txt";
+  se_path = wcsimdir + "second_excited_spectrum.txt";
+
+  this->Initialise();
+}
+
+WCSimAmBeGen::~WCSimAmBeGen(){
+  
+  // This needed to be deleted
+  delete myAmBeGun;
+  delete rGen;
+  delete nEnergyDist;
+}
+
+void WCSimAmBeGen::Initialise(){
+    nEnergyDist = new G4SPSEneDistribution();
+
+    vtx = G4ThreeVector(0., 0., 0.);
+    time = 0.;
+    epsilon = 1e-6;
+}
+
+G4double WCSimAmBeGen::GammaEnergy(){
+    
+    // Initialise Geant4 random number generator
+    G4Random::setTheEngine(new CLHEP::RanecuEngine);
+
+    // Generation of the random number
+    G4double random_number = G4UniformRand();
+
+    // Select the energy taking into account the probabilities
+    G4double cumulative_probability = 0.0;
+    G4int selected_energy_index = -1;
+
+    for (G4int j = 0; j < nGammaOutcomes; ++j) {
+        cumulative_probability += gammaProbabilities[j];
+        if (random_number <= cumulative_probability) {
+            selected_energy_index = j;
+            break;
+        }
+    }
+
+    gEnergy = gammaEnergies[selected_energy_index];
+
+    return gEnergy; 
+}
+
+G4double WCSimAmBeGen::NeutronEnergy(G4double gamEnergy){
+    nEnergyDist->SetEnergyDisType("Arb"); 
+ 
+    // Depending on the gEnergy, we load the correspondent neutron energy spectrum 
+    if (std::abs(gEnergy - 0.0) < epsilon) { 
+      nEnergyDist->ArbEnergyHistoFile(gs_path);
+      nEnergyDist->SetBiasRndm(rGen);
+      nEnergyDist->ArbInterpolate("Lin");
+      nEnergy = nEnergyDist->GenerateOne(G4Neutron::Definition());
+    }
+    else if (std::abs(gEnergy - 4.4) < epsilon) {
+      nEnergyDist->ArbEnergyHistoFile(fe_path);
+      nEnergyDist->SetBiasRndm(rGen);
+      nEnergyDist->ArbInterpolate("Lin");
+      nEnergy = nEnergyDist->GenerateOne(G4Neutron::Definition());
+    }
+    else {
+      nEnergyDist->ArbEnergyHistoFile(se_path);
+      nEnergyDist->SetBiasRndm(rGen);
+      nEnergyDist->ArbInterpolate("Lin");
+      nEnergy = nEnergyDist->GenerateOne(G4Neutron::Definition());
+    }
+
+    return nEnergy;
+}
+
+void WCSimAmBeGen::GenerateNG(G4Event* anEvent){
+    for (G4int i=0; i<2; i++){
+      // Create a vertex for the particle following the index of the loop
+      G4int pdgid = pdgids[i];
+      
+      // Configure particle's properties in particleGun
+      if (pdgid == 22) {
+        myAmBeGun->SetParticleDefinition(G4Gamma::Definition());
+        myAmBeGun->SetParticleEnergy(GammaEnergy());
+      }
+      else {
+        myAmBeGun->SetParticleDefinition(G4Neutron::Definition());
+        myAmBeGun->SetParticleEnergy(NeutronEnergy(GammaEnergy()));
+      }
```

Comment:
```suggestion
      if(i == 0)
        GammaEnergy();
      if (pdgid == 22) {
        myAmBeGun->SetParticleDefinition(G4Gamma::Definition());
        myAmBeGun->SetParticleEnergy(gEnergy);
      }
      else {
        myAmBeGun->SetParticleDefinition(G4Neutron::Definition());
        myAmBeGun->SetParticleEnergy(NeutronEnergy());
      }
```
This will ensure the neutron & photon generated in each event is correctly correlated

Note that this should be `pdgids` order-independent